SConstruct

```diff
@@ -546,6 +546,9 @@ else:
 		env.Append(
 			CXXFLAGS = [
 				"/WX",  # Treat warnings as errors
+				# We are building all client code in the exact same environment, so we can safely
+				# disable warnings about not exporting private classes
+				"/wd4251",
```

Comment:
It does seem like it could be a trickier situation than just turning off the warning, and maybe not necessarily as simple as it showing up as a linker warning in problematic cases. But I don't really have a better solution at the moment. A deduplicated list of this error turns up 310 instances where it's happening, many from STL, boost or OpenVDB members of a class.

Based on [one explanation](https://stackoverflow.com/questions/5661738/how-can-i-use-standard-library-stl-classes-in-my-dll-interface-or-abi), it sounds like due to the nature of STL being a standard rather than a concrete implementation, the ABI is not _necessarily_ stable across versions or compilers.

A few solutions has been [put forward](https://stackoverflow.com/questions/16419318/one-way-of-eliminating-c4251-warning-when-using-stl-classes-in-the-dll-interface), but they all seem to have pretty significant drawbacks.

Given all that, I come down on the side of requiring Gaffer extensions on Windows to be built with the same MSVC version we use. They don't get updated all that frequently, every two or three years lately, and at least for the 2017-2019-2022 chain they claim to be binary compatible.

Does that sound reasonable?

Reply:
I clarified the comment about disabling the warning in https://github.com/GafferHQ/gaffer/pull/5277/commits/7119faf68d18ced2e7c6f510ed3e7a3397ce3674

If we agree to go this route, is there a good place to add a note to downstream folks who are building extensions?

Reply:
Yep, I think it's completely reasonable to require extension authors to use a specific MSVC version - that's normal for any DCC.

I wasn't clear enough in my original comment. What I was thinking of is the case where we've forgotten to DLLEXPORT something in the API that an extension author might want to use. In that case I think they'll get a linking error if they try to use it. But because we've silenced the warning, we won't know about it until then. Does that make sense?

In any case, I still think silencing the warning is the best compromise.

> If we agree to go this route, is there a good place to add a note to downstream folks who are building extensions?

Not that I can think of, in the absence of extension authoring tutorials in the docs.